### DISCUSSION TITLE
Sending commands to drone in the X-Plane simulator with inav SITL

### DISCUSSION THREAD

**Discussion by NikitaYeryomin** [iNavFlight/inav#9246] [General]
Hello INAV community,

I'm currently in the process of developing an autopilot system, with the decision-making software running on a Jetson Nano. My primary objective is to have the drone react in a simulator based on commands sent from my script.
I'm using the following setup:

1. X-Plane 11 simulator from Steam on Windows 10 with NK_FPVSurfwing model from the [INAV-X-Plane-HITL](https://github.com/RomanLut/INAV-X-Plane-HITL) repo. I have configured Data output to the 49000 port. Settings can be viewed below: ![x-plane data output](https://res.cloudinary.com/disqgapnj/image/upload/v1692259009/xplane_data_output1_grlwba.jpg)
I've also enabled all [datarefs](https://res.cloudinary.com/disqgapnj/image/upload/v1692259009/xplane_data_output2_l7vfbs.jpg) I found in the [xplane.c](https://github.com/iNavFlight/inav/blob/master/src/main/target/SITL/sim/xplane.c) code of the inav SITL.

2. inav_SITL running on the same host in WSL (Ubuntu 20.04) with the following command line:
`inav_SITL --sim=xp --simip=192.168.0.162 --simport=49000 --chanmap=S01-01,S02-02,S03-03,M01-04`
It's able to successfully connect to X-Plane: ![screenshot](https://res.cloudinary.com/disqgapnj/image/upload/v1692259899/inav_sitl_uloir5.jpg)

3. INAV Configurator 6.1.0 - successfully connects the inav_SITL on port 5760 and I can see it receives the data updates from the simulator: ![screenshot](https://res.cloudinary.com/disqgapnj/image/upload/v1692260317/inav_conf_setup_uwypu8.jpg)
Here are also my ports and receiver configuration tabs: 
![ports](https://res.cloudinary.com/disqgapnj/image/upload/v1692260317/inav_conf_ports_glproq.jpg)
![receiver](https://res.cloudinary.com/disqgapnj/image/upload/v1692260317/inav_conf_receiver_zd3lba.jpg)
 
To communicate with the inav_SITL with my script I've created [this fork](https://github.com/NikitaYeryomin/pyMultiWii) of pyMultiWii library. It uses TCP socket instead of serial to connect to inav_SITL and send MSP commands. I'm able to successfully connect to inav_SITL on port 5761 and retrieve any data like attitude or RC data, e.g.:

``````from pymultiwii import MultiWii
import time

def get_rc_data(board):
    rc_data = board.getData(MultiWii.RC)
    return rc_data

if __name__ == "__main__":
    host = '127.0.0.1'
    port = 5761
    board = MultiWii(host, port)
    
    while True:
        try:
            rc_data = get_rc_data(board)
            if rc_data is not None:
                print(rc_data)
            else:
                print("Could not retrieve RC data.")
        except Exception as e:
           print(f"Error: {e}")
        time.sleep(0.1)
``````

    However, I'm not able to send any commands to the drone, like arming or changing throttle. Here's the code I've tried to use for take off:

``````
def take_off(drone, takeoff_throttle, duration):
    if not drone.arm():
        print("Failed to arm the drone!")
        return

    print("Taking off...")
    start_time = time.time()

    hover_throttle = 1500  
    elevon_neutral = 1500 

    start_time = time.time()
    while time.time() - start_time < duration:
        # Set throttle for takeoff and maintain elevon in neutral position
        data = [elevon_neutral, elevon_neutral, 1500, takeoff_throttle]
        drone.sendCMD(8, MultiWii.SET_RAW_RC, data, '4H')
        time.sleep(0.1)

    # Set throttle to hover after takeoff
    data = [elevon_neutral, elevon_neutral, 1500, hover_throttle]
    drone.sendCMD(8, MultiWii.SET_RAW_RC, data, '4H')
    print("Hovering...")
``````

    It does not respond with any error but looks like it just ignores them as I don't see any visual changes in the simulator as well as it does not affect any RC data values. 

Has anyone faced a similar issue or can provide guidance on what might be going wrong?

Any help or insight much appreciated!

**Comment by stronnag** [comment]
In principle, sending MSP to the SITL works fine, for example [fl2stil](https://github.com/stronnag/bbl2kml/wiki/fl2sitl) provides RC commands via `MSP_SET_RAW_RC` in order to replay BBL. 

There is also further information on [automation in the wiki](https://github.com/iNavFlight/inav/wiki/INAV-Remote-Management,-Control-and-Telemetry).

In particular for arming and sending RC commands there are specific status requirements and update rate requirements.

**Comment by pogopo789** [comment]
try to change from SIM (SITL) to MSP
![68747470733a2f2f7265732e636c6f7564696e6172792e636f6d2f646973716761706e6a2f696d6167652f75706c6f61642f76313639323236303331372f696e61765f636f6e665f72656365697665725f7a64336c62612e6a7067](https://github.com/iNavFlight/inav/assets/46533062/972c9ae6-8858-4b45-974d-fe7b3545adf6)

**Comment by walkvanja** [comment]
@NikitaYeryomin have you resolved an issue? Have the same and looking for the solution. Approach from @pogopo789 change stick values in iNav configurator, but not affect simulation in RealFlight

Any help would be appreciated

**Comment by Badaszz** [comment]
how did you set up inav sitl, i am having problems with it
